core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -3435,6 +3437,43 @@ object SparkContext extends Logging {
     supplement(DRIVER_JAVA_OPTIONS)
     supplement(EXECUTOR_JAVA_OPTIONS)
   }
+
+  private def supplementJavaGCOptions(conf: SparkConf): Unit = {
+    val javaFlagCommand = "java -XX:+PrintFlagsFinal -version"
+    val gcAlgorithmPattern = """^\s*bool\s+(Use\w+GC)\s*[=:].*""".r
+    // These flags contain "GC" in their names but aren't standalone garbage collection algorithms.
+    // They're specific policy options that modify GC behavior rather than representing
+    // complete GC implementations, so we exclude them when detecting available GC algorithms.
+    val nonGCAlgorithmFlags: Set[String] = Set(
+      "UseAdaptiveSizePolicyWithSystemGC",
+      "UseMaximumCompactionOnSystemGC")
+
+    val gcAlgorithms = try {
+      val output = javaFlagCommand.!!
```

Comment:
Oh, is this a way to detect `JDK_JAVA_OPTIONS` too, @wangyum ? I was curious how we can handle the user environment.

Reply:
The new implementation checks if a GC algorithm is specified in the configuration: https://github.com/apache/spark/commit/38eacf02d8e9c1fddcfa2b3f1566e0e6e85673d4